modules/ding_innskraning/ding_innskraning.module

```diff
@@ -0,0 +1,780 @@
+}
+
+/**
+ * Consume authentications from https://innskraning.island.is.
+ */
+function ding_innskraning_saml_endpoint() {
+  // Extract token from request.
+  if (!array_key_exists('token', $_POST)) {
+    watchdog('ding_innskraning', "Post parameter 'token' not found in SAML response.", array(), WATCHDOG_WARNING);
+    drupal_goto('user/login');
+    return FALSE;
+  }
+  $token = $_POST['token'];
+  // Decode the request.
+  $xml = base64_decode($token);
+
+  // Load XML document.
+  $dom = new DOMDocument();
+  $dom->loadXML($xml);
+
+  if (!$dom) {
+    drupal_set_message(t('Authentication failed'), 'error');
+    watchdog('ding_innskraning', 'SAML Response could not be processed, XML: @xml', array('@xml' => $xml), WATCHDOG_ERROR);
+    drupal_goto('user/login');
+    return FALSE;
+  }
+
+  // Retrieve name.
+  $name = _ding_innskraning_get_attribute_by_name($dom, 'Name');
+  // Retrieve SSN.
+  $ssn = _ding_innskraning_get_attribute_by_name($dom, 'UserSSN');
+  // Provide a message.
+  watchdog('ding_innskraning', "Attempt to login %name", array('%name' => $name), WATCHDOG_DEBUG);
+  // Verify response.
+  try {
+    _ding_innskraning_verify($dom);
+  }
+  catch (Exception $ex) {
+    drupal_set_message(t('Authentication failed'), 'error');
+    watchdog_exception(
+      'ding_innskraning',
+      $ex,
+      'Error verifying token, %type: !message in %function (line %line of %file), XML: @xml',
+      array('@xml' => $xml),
+      WATCHDOG_ERROR
+    );
+    drupal_goto('user/login');
+  }
+
+  // Make sure user is not logged ind.
+  $login_success = FALSE;
+  if (user_is_logged_in()) {
+    drupal_set_message(t('You cannot login with https://innskraning.island.is while you are logged in. Please log out first'), 'error');
+  }
+  else {
+    // OK, let's log the user in.
+    $login_success = _ding_innskraning_activate_user_by_ssn($ssn);
+  }
+
+  if ($login_success) {
+    // Go to frontpage.
+    drupal_goto('<front>');
+  }
+  else {
+    // Go to login page.
+    drupal_set_message(
+      t('Authentication through @link failed! Authentication through @link only available to registered users.',
+        array('@link' => l(t('island.is'), 'https://island.is'))
+      ),
+      'error'
+    );
+    drupal_goto('user/login');
+  }
+}
+
+/**
+ * Generate a SAML auth id.
+ *
+ * Creates an id for the authentication request. island.is requires the id to be
+ * formatted as a GUID.
+ */
+function _ding_innskraning_auth_id() {
+  // Implementation created from
+  // https://www.texelate.co.uk/blog/create-a-guid-with-php
+  // Create a token.
+  $token = $_SERVER['HTTP_HOST'];
+  $token .= $_SERVER['REQUEST_URI'];
+  $token .= uniqid(rand(), TRUE);
+
+  // GUID is 128-bit hex.
+  $hash = strtoupper(md5($token));
+
+  // Create formatted GUID.
+  $guid = '';
+
+  // GUID format is XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX for readability.
+  $guid .= substr($hash, 0, 8) . '-' . substr($hash, 8, 4) . '-' . substr($hash, 12, 4) .
```

Comment:
Why append to an empty string instead of just setting it?

Reply:
Fixed